src/catch2/internal/catch_output_redirect.cpp

```diff
@@ -60,85 +63,202 @@ namespace Catch {
-         // We manually create the file on Windows only, on Linux
-         // it will be autodeleted
+    if (result)
+    {
+        throw std::system_error{ errno, std::generic_category() };
+    }
+}
+
+inline int read_or_throw(int descriptor, void* buffer, size_t size)
+{
 #if defined(_MSC_VER)
-         std::remove(m_buffer);
+    int result{ _read(descriptor, buffer, static_cast<unsigned>(size)) };
+#else
+    int result{ read(descriptor, buffer, size) };
 #endif
+
+    if (result == -1)
+    {
+        throw std::system_error{ errno, std::generic_category() };
     }
 
+    return result;
+}
 
-    FILE* TempFile::getFile() {
-        return m_file;
+inline void fflush_or_throw(std::FILE* file)
+{
+    if (std::fflush(file))
+    {
+        throw std::system_error{ errno, std::generic_category() };
     }
+}
 
-    std::string TempFile::getContents() {
-        std::stringstream sstr;
-        char buffer[100] = {};
-        std::rewind(m_file);
-        while (std::fgets(buffer, sizeof(buffer), m_file)) {
-            sstr << buffer;
-        }
-        return sstr.str();
+jthread::jthread() noexcept : m_thread{} {}
+
+template <typename F, typename... Args>
+jthread::jthread(F&& f, Args&&... args) : m_thread{ std::forward<F>(f), std::forward<Args>(args)... } {}
+
+// Not exactly like std::jthread, but close enough for the code below.
+jthread::~jthread() noexcept
+{
+    if (m_thread.joinable())
+    {
+        m_thread.join();
     }
+}
+
+constexpr UniqueFileDescriptor::UniqueFileDescriptor() noexcept : m_value{} {}
 
-    OutputRedirect::OutputRedirect(std::string& stdout_dest, std::string& stderr_dest) :
-        m_originalStdout(dup(1)),
-        m_originalStderr(dup(2)),
-        m_stdoutDest(stdout_dest),
-        m_stderrDest(stderr_dest) {
-        dup2(fileno(m_stdoutFile.getFile()), 1);
-        dup2(fileno(m_stderrFile.getFile()), 2);
+UniqueFileDescriptor::UniqueFileDescriptor(int value) noexcept : m_value{ value } {}
+
+constexpr UniqueFileDescriptor::UniqueFileDescriptor(UniqueFileDescriptor&& other) noexcept :
+    m_value{ other.m_value }
+{
+    other.m_value = 0;
+}
+
+UniqueFileDescriptor::~UniqueFileDescriptor() noexcept
+{
+    if (m_value == 0)
+    {
+        return;
     }
 
-    OutputRedirect::~OutputRedirect() {
-        Catch::cout() << std::flush;
-        fflush(stdout);
-        // Since we support overriding these streams, we flush cerr
-        // even though std::cerr is unbuffered
-        Catch::cerr() << std::flush;
-        Catch::clog() << std::flush;
-        fflush(stderr);
+    close_or_throw(m_value); // std::terminate on failure (due to noexcept)
+}
 
-        dup2(m_originalStdout, 1);
-        dup2(m_originalStderr, 2);
+UniqueFileDescriptor& UniqueFileDescriptor::operator=(UniqueFileDescriptor&& other) noexcept
+{
+    if (this != &other)
+    {
+        if (m_value != 0)
+        {
+            close_or_throw(m_value); // std::terminate on failure (due to noexcept)
```

Comment:
The code before had a TBD here and just ignored failures; let me know if you prefer that approach. (I think std::terminate or ignore are the two most obvious options here.)